src/macros.rs

```diff
@@ -0,0 +1,336 @@
+/// Construct a `scale_value::Value`
+///
+///
+/// Supports unnamed and named composites and variants:
+/// ```
+/// use scale_value::value;
+///
+/// let val = value!({
+///     name: "localhost",
+///     address: V4(127, 0, 0, 1),
+///     payload: {
+///         bytes: (255, 3, 4, 9),
+///         method: ("Post", 3000),
+///     },
+/// });
+///
+/// ```
+/// Values can be nested in each other:
+/// ```
+/// use scale_value::value;
+///
+/// let data_value = value!((1, v1(1, 2), 3));
+/// let val = value!(POST { data: data_value });
+/// ```
+/// Trailing commas are optional.
+#[macro_export(local_inner_macros)]
+macro_rules! value {
+    ($($tt:tt)*) => {
+        value_internal!($($tt)*)
+    };
+}
+
+/// All patterns can be grouped into 4 main patterns:
+///
+/// ### `value_internal!(@unnamed [ELEMENTS] (REST))`
+///
+/// checks `REST` for certain patterns and if a suitable element pattern is found,
+/// the element is added to the comma seperated `ELEMENTS` list. When `REST` is empty,
+/// we collect all `ELEMENTS` into a Vec of values.
+///
+/// ### `value_internal!(@named [KEYVALUEPAIRS] (PARTIALKEY) (REST))`
+///
+/// goes over the REST tokens, to build up the PARTIALKEY until it is a proper KEY.
+/// This happens as soon as a colon `:` token is encountered, then we switch to the next pattern:
+///
+/// ### `value_internal!(@named [KEYVALUEPAIRS] [KEY] (REST))`
+///
+/// The square brackets now indicate that the key is fully formed. Now REST is scanned for the next
+/// `VALUE`. Together with the `KEY`, they are added as a key value pair tuple into the KEYVALUEPAIRS list.
+/// At that point we switch back to the partial key pattern above, e.g. `value_internal!(@named [KEYVALUEPAIRS] () (REST))`
+/// now with a new empty partial key that has to be filled.
+/// When `REST` is empty, we collect all `KEYVALUEPAIRS` into a Vec of key-value tuples.
+///
+/// ### `value_internal!(BASEPATTERN)`
+///
+/// These patterns check if the input represents a composite or variant type or can be made into a valid `$crate::Value`.
+#[macro_export(local_inner_macros)]
+#[doc(hidden)]
+macro_rules! value_internal {
+
+    ////////////////////////////////////////////////////////////////////////////
+    // collecting unnamed fields
+    ////////////////////////////////////////////////////////////////////////////
+
+    // done, put all the fields in a vec
+    (@unnamed [$($e:expr, )*] ()) => { vec_wrapper![$($e, )*] };
```

Comment:
Dumb question: Don't we need another arm to match the expression without ending comma? 

```rust
    (@unnamed [$($e:expr ), *] ()) => { vec_wrapper![$($e, )*] };
```

Reply:
If there is no trailing comma, the arm on line 101 should catch it instead and package it into a vec: https://github.com/paritytech/scale-value/blob/tadeo-hepperle-value-macro/src/macros.rs#L101